lib/data/kinetic/Kinetic.js

```diff
@@ -0,0 +1,302 @@
+import stream from 'stream';
+
+import { errors } from 'arsenal';
+import kineticlib from 'kineticlib';
+
+const HEADER_SZ = 9;
+
+class Kinetic {
+
+    constructor(drives) {
+        this._events = new EventEmitter();
+        this._sockets = [];
+        this._drives = drives;
+        for (let i = 0; i < drives.length; i++) {
+            this._sockets.push(
+                { sock: this._newSocket(),
+                  connectionID: 0,
+                  clusterVersion: 0,
+                  drive: drives[i] });
+        }
+        this._request = false;
+        this._chunkSize = 0;
+        this._chunkTab = [];
+        this._count = 0;
+        this.connect();
+        return this;
+    }
+
+    connect() {
+        let index = 0;
+        this._sockets.forEach(socket => {
+            socket.sock.on('connect', () => {
+                this._bindReadable(socket, index);
+                ++index;
+            });
+            socket.sock.connect(socket.drive);
+        });
+    }
+
+    _bindReadable(socket, index) {
+        socket.sock.on('readable', err => {
+            if (this._request === false) {
+                this._parsePDU(socket, (err1, pdu) => {
+                    this._parseRequest(socket, err || err1, pdu, index);
+                });
+            } else {
+                this._endGet(socket);
+            }
+        });
+    }
+
+    _parsePDU(socket, callback) {
+        const header = socket.sock.read(HEADER_SZ);
+        const protobufSize = header.readInt32BE(1);
+        const rawData = socket.sock.read(protobufSize);
+        const pdu = new kineticlib.PDU(Buffer.concat([header, rawData]));
+        const err = this._propError(pdu);
+        return callback(err, pdu);
+    }
+
+    _parseRequest(socket, err, pdu, index) {
+        const sock = socket;
+        switch (pdu.getMessageType()) {
+        case null:
+            this._initPDU = pdu;
+            sock.connectionID = pdu.getConnectionId();
+            sock.clusterVersion = pdu.getClusterVersion();
+            sock.index = index;
+            sock.sequence = 0;
+            return pdu;
+        case kineticlib.ops.PUT_RESPONSE:
+            this._events.emit('putResponse', err);
+            break;
+        case kineticlib.ops.DELETE_RESPONSE:
+            this._events.emit('deleteResponse', err);
+            break;
+        case kineticlib.ops.GETLOG_RESPONSE:
+            this._events.emit(
+                'getLogResponse', err, socket, pdu.getLogObject());
+            break;
+        case kineticlib.ops.GET_RESPONSE:
+            if (err) {
+                this._events.emit('getResponse', err);
+            } else {
+                this._startGet(socket, pdu);
+            }
+            break;
+        default:
+            break;
+        }
+        return undefined;
+    }
+
+    _startGet(socket, pdu) {
+        let chunk = Buffer.allocUnsafe(0);
+        this._request = true;
+        this._chunkSize = pdu.getChunkSize();
```

Comment:
In the case of multiple `get`s of different-size objects on a same socket, `this._chunkSize` will be not correct.